### PR TITLE
Add a script to list the routes in a region

### PR DESCRIPTION
**Pull request by richardTowers**
What
----

This may be useful when we're trying to understand what things are
running on the platform.

Possible future improvements:

- send a `dig` to custom domains to see if DNS is actually setup
- send an HTTP request to the domain to see if it's serving 200s

[skip-ci]

How to review
-------------

* Code review
* Run the script yourself and check it works

Who can review
--------------

Not @richardtowers

### CODE REVIEW

File: scripts/list-routes.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+require 'json'
+require 'set'
+
+routes_by_domain_url = {}
+
+page = JSON.load `cf curl '/v2/routes?results-per-page=100'`
+next_url = page["next_url"]
+
+while next_url
+  page = JSON.load `cf curl '#{next_url}'`
```
Comment (46bit) [inline scripts/list-routes.rb]: This reads like it accidentally discards the first page of results. Setting `next_url = '/v2/routes?results-per-page=100'` would workaround that.

Comment (richardTowers) [reply scripts/list-routes.rb]: Good catch - I'm just making deliberate mistakes to keep you on your toes. 🤦‍♂

Comment (46bit) [reply scripts/list-routes.rb]: It's always nice to know that code review can catch bugs :)

Comment (46bit) [reply scripts/list-routes.rb]: I'll re-review after lunch 🥖

File: scripts/list-routes.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+require 'set'
+
+routes_by_domain_url = {}
+
+next_url = '/v2/routes?results-per-page=100'
+while next_url
+  page = JSON.parse `cf curl '#{next_url}'`
+  page['resources'].map do |resource|
+    host = resource.dig('entity', 'host')
+    unless host.empty?
```
Comment (46bit) [inline scripts/list-routes.rb]: In London this excludes 51 routes which do have an empty `host` field. Most seem to be using path-based routing. 😞

Comment (46bit) [reply scripts/list-routes.rb]:
```suggestion
```

File: scripts/list-routes.rb [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+      domain_url = resource.dig('entity', 'domain_url')
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    domain_url = resource.dig('entity', 'domain_url')
```

File: scripts/list-routes.rb [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      if routes_by_domain_url.include? domain_url
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    if routes_by_domain_url.include? domain_url
```

File: scripts/list-routes.rb [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+        routes_by_domain_url[domain_url] << host
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
      routes_by_domain_url[domain_url] << host
```

File: scripts/list-routes.rb [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+      else
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    else
```

File: scripts/list-routes.rb [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        routes_by_domain_url[domain_url] = [host].to_set
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
      routes_by_domain_url[domain_url] = [host].to_set
```

File: scripts/list-routes.rb [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      end
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    end
```

File: scripts/list-routes.rb [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+    end
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
```

File: scripts/list-routes.rb [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+  end
+  next_url = page["next_url"]
+end
+
+routes_by_domain_url.each do |url, hosts|
+  domain = JSON.parse(`cf curl '#{url}'`).dig('entity', 'name')
+  hosts.each do |host|
+    puts "https://#{host}.#{domain}"
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    fqdn = host.empty? ? "#{domain}" : "#{host}.#{domain}"
    puts "https://#{fqdn}"
```

File: scripts/list-routes.rb [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+      routes_by_domain_url[domain_url] = [host].to_set
+    end
+  end
+  next_url = page["next_url"]
+end
+
+routes_by_domain_url.each do |url, hosts|
+  domain = JSON.parse(`cf curl '#{url}'`).dig('entity', 'name')
+  hosts.each do |host|
+    fqdn = host.empty? ? "#{domain}" : "#{host}.#{domain}"
```
Comment (46bit) [inline scripts/list-routes.rb]:
```suggestion
    fqdn = host.empty? ? domain : "#{host}.#{domain}"
```

**Pull request merged**